File: prototype/src/shalom_gcs/widgets/gauges.py
"""QPainter 로 직접 그리는 계측 위젯.

기본 위젯(QProgressBar 등)에 QSS 를 발라도 계측기처럼 보이지 않는다.
반면 안티에일리어싱 호/막대는 QPainter 로 짧게 나오고, 이게 화면 인상의
대부분을 만든다. 웹 대비 Qt 가 유리한 지점.

값 변화는 QPropertyAnimation 으로 이징한다. 다만 짧게(280ms) 둔다 —
관제 화면에서 애니메이션이 길면 '지금 값'을 못 믿게 된다.

색상은 paint 시점에 palette() 를 조회한다. 재시작 없이 테마가 바뀌게 하기 위함.
"""

from __future__ import annotations

import math

from PySide6.QtCore import Property, QEasingCurve, QPropertyAnimation, QRectF, Qt
from PySide6.QtGui import QColor, QFont, QPainter, QPen
from PySide6.QtWidgets import QSizePolicy, QWidget

from ..theme.tokens import mono_family, palette


class _Animated(QWidget):
    """부드럽게 따라가는 스칼라 값 하나를 갖는 베이스."""

    def __init__(self, parent: QWidget | None = None, *, duration: int = 280):
        super().__init__(parent)
        self._value = 0.0
        self._anim = QPropertyAnimation(self, b"value", self)
        self._anim.setDuration(duration)
        self._anim.setEasingCurve(QEasingCurve.OutCubic)

    def get_value(self) -> float:
        return self._value

    def set_value(self, v: float) -> None:
        self._value = float(v)
        self.update()

    value = Property(float, get_value, set_value)

    def animate_to(self, target: float) -> None:
        if abs(target - self._value) < 1e-4:
            return
        self._anim.stop()
        self._anim.setStartValue(self._value)
        self._anim.setEndValue(float(target))
        self._anim.start()


def _mono(size: int, bold: bool = False) -> QFont:
    f = QFont(mono_family())
    f.setPointSize(size)
    if bold:
        f.setWeight(QFont.DemiBold)
    return f


class BatteryRing(_Animated):
    """배터리 잔량 원형 게이지.

    시나리오 예외처리에 '최소 배터리 용량 미만 시 즉시 충전 복귀'가 있어
    임계값 위치를 눈금으로 명시한다.
    """

    def __init__(self, parent: QWidget | None = None, *, size: int = 86,
                 low_threshold: float = 25.0):
        super().__init__(parent)
        self._size = size
        self._low = low_threshold
        self._charging = False
        self.setFixedSize(size, size)

    def set_state(self, soc_pct: float, charging: bool = False) -> None:
        self._charging = charging
        self.animate_to(max(0.0, min(100.0, soc_pct)))

    def _arc_color(self) -> QColor:
        P = palette()
        if self._charging:
            return QColor(P.accent)
        if self._value <= self._low * 0.6:
            return QColor(P.danger)
        if self._value <= self._low:
            return QColor(P.warning)
        return QColor(P.success)

    def paintEvent(self, _ev) -> None:
        P = palette()
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)

        thickness = 6.0
        pad = thickness / 2 + 1
        rect = QRectF(pad, pad, self.width() - pad * 2, self.height() - pad * 2)

        p.setPen(QPen(QColor(P.surface_hi if P.is_dark else P.surface_hover),
                      thickness, Qt.SolidLine, Qt.FlatCap))
        p.drawArc(rect, 0, 360 * 16)

        if self._value > 0.3:
            p.setPen(QPen(self._arc_color(), thickness, Qt.SolidLine, Qt.FlatCap))
            p.drawArc(rect, 90 * 16, -int(360 * 16 * self._value / 100.0))

        # 임계값 눈금 — 배경색으로 호를 끊어 표시
        ang = math.radians(90 - 360 * self._low / 100.0)
        cx, cy = self.width() / 2, self.height() / 2
        r_out = rect.width() / 2 + thickness / 2
        r_in = rect.width() / 2 - thickness / 2
        p.setPen(QPen(QColor(P.surface), 2))
        p.drawLine(cx + r_in * math.cos(ang), cy - r_in * math.sin(ang),
                   cx + r_out * math.cos(ang), cy - r_out * math.sin(ang))

        p.setFont(_mono(int(self._size * 0.21), bold=True))
        p.setPen(QColor(P.text))
        p.drawText(self.rect().adjusted(0, -5, 0, -5), Qt.AlignCenter, f"{self._value:.0f}")

        f2 = QFont(); f2.setPointSize(max(7, int(self._size * 0.09)))
        p.setFont(f2)
        p.setPen(QColor(P.text_mute))
        off = int(self._size * 0.27)
        p.drawText(self.rect().adjusted(0, off, 0, off), Qt.AlignCenter,
                   "충전 중" if self._charging else "%")
        p.end()


class ArcGauge(_Animated):
    """반원 호 게이지 — FR3 조작성 지수(특이자세 근접도).

    낮을수록 위험한 지표라 색이 역방향이다. 실제 특이자세 판정·회피는
    로봇측 암 노드 책임이고 여기선 표시만 한다 (bridge_protocol.md §4).
    """

    def __init__(self, parent: QWidget | None = None, *, caption: str = "",
                 warn_below: float = 0.35, danger_below: float = 0.15):
        super().__init__(parent)
        self._caption = caption
        self._warn = warn_below
        self._danger = danger_below
        self._raw_text = "—"
        self.setMinimumHeight(70)
        self.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Fixed)

    def set_state(self, normalized: float, raw_text: str = "") -> None:
        self._raw_text = raw_text or f"{normalized:.3f}"
        self.animate_to(max(0.0, min(1.0, normalized)))

    def tone_color(self) -> QColor:
        P = palette()
        if self._value <= self._danger:
            return QColor(P.danger)
        if self._value <= self._warn:
            return QColor(P.warning)
        return QColor(P.success)

    def paintEvent(self, _ev) -> None:
        P = palette()
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)

        thickness = 6.0
        w, h = self.width(), self.height()
        d = min(w - 14, (h - 22) * 2)
        rect = QRectF((w - d) / 2, 9, d, d)

        p.setPen(QPen(QColor(P.surface_hi if P.is_dark else P.surface_hover),
                      thickness, Qt.SolidLine, Qt.FlatCap))
        p.drawArc(rect, 180 * 16, -180 * 16)

        p.setPen(QPen(self.tone_color(), thickness, Qt.SolidLine, Qt.FlatCap))
        p.drawArc(rect, 180 * 16, -int(180 * 16 * self._value))

        p.setFont(_mono(12, bold=True))
        p.setPen(QColor(P.text))
        p.drawText(QRectF(0, h - 32, w, 18), Qt.AlignCenter, self._raw_text)

        if self._caption:
            f2 = QFont(); f2.setPointSize(8)
            p.setFont(f2)
            p.setPen(QColor(P.text_mute))
            p.drawText(QRectF(0, h - 15, w, 13), Qt.AlignCenter, self._caption)
        p.end()


class StatBar(QWidget):
    """'라벨 ─ 값' 한 줄 막대. CPU/GPU 온도, 사용률, 링크 지연."""

    def __init__(self, label: str, unit: str = "", parent: QWidget | None = None,
                 *, warn_above: float | None = None, danger_above: float | None = None,
                 vmax: float = 100.0):
        super().__init__(parent)
        self._label = label
        self._unit = unit
        self._warn = warn_above
        self._danger = danger_above
        self._vmax = vmax
        self._value = 0.0
        self._valid = False
        self.setFixedHeight(26)

    def set_value(self, v: float | None) -> None:
        self._valid = v is not None
        self._value = float(v or 0.0)
        self.update()

    def _color(self) -> QColor:
        P = palette()
        if not self._valid:
            return QColor(P.text_mute)
        if self._danger is not None and self._value >= self._danger:
            return QColor(P.danger)
        if self._warn is not None and self._value >= self._warn:
            return QColor(P.warning)
        return QColor(P.text_dim)

    def paintEvent(self, _ev) -> None:
        P = palette()
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)
        w = self.width()

        f = QFont(); f.setPointSize(9)
        p.setFont(f)
        p.setPen(QColor(P.text_mute))
        p.drawText(QRectF(0, 0, w * 0.55, 13), Qt.AlignLeft | Qt.AlignVCenter, self._label)

        p.setFont(_mono(9))
        p.setPen(QColor(P.text) if self._valid else QColor(P.text_mute))
        txt = f"{self._value:.0f}{self._unit}" if self._valid else "—"
        p.drawText(QRectF(w * 0.45, 0, w * 0.55, 13), Qt.AlignRight | Qt.AlignVCenter, txt)

        p.setPen(Qt.NoPen)
        p.setBrush(QColor(P.surface_hi if P.is_dark else P.surface_hover))
        p.drawRect(QRectF(0, 18, w, 3))

        if self._valid and self._value > 0:
            frac = max(0.0, min(1.0, self._value / self._vmax))
            p.setBrush(self._color())
            p.drawRect(QRectF(0, 18, max(2.0, w * frac), 3))
        p.end()


class JointBar(QWidget):
    """관절 하나의 실제값을 한계 범위 안에서 표시.

    조작은 QSlider 가 하고 이건 표시 전용이다. 명령값과 실제값이 갈라지는
    상황(계획 중, 리플렉스 중단)을 조작자가 봐야 하므로 둘을 함께 그린다.
    """

    def __init__(self, name: str, lo: float, hi: float, parent: QWidget | None = None):
        super().__init__(parent)
        self._name = name
        self._lo, self._hi = lo, hi
        self._actual = 0.0
        self._command: float | None = None
        self.setFixedHeight(22)

    def set_actual(self, rad: float) -> None:
        self._actual = rad
        self.update()

    def set_command(self, rad: float | None) -> None:
        self._command = rad
        self.update()

    def _frac(self, v: float) -> float:
        return max(0.0, min(1.0, (v - self._lo) / (self._hi - self._lo)))

    def _near_limit(self) -> bool:
        f = self._frac(self._actual)
        return f < 0.04 or f > 0.96

    def paintEvent(self, _ev) -> None:
        P = palette()
        p = QPainter(self)
        p.setRenderHint(QPainter.Antialiasing)
        w, h = self.width(), self.height()
        bar_x = 26
        bar_w = w - bar_x - 56

        f = QFont(); f.setPointSize(9)
        p.setFont(f)
        p.setPen(QColor(P.text_mute))
        p.drawText(QRectF(0, 0, 22, h), Qt.AlignLeft | Qt.AlignVCenter, self._name)

        y = h / 2 - 1.5
        p.setPen(Qt.NoPen)
        p.setBrush(QColor(P.surface_hi if P.is_dark else P.surface_hover))
        p.drawRect(QRectF(bar_x, y, bar_w, 3))

        # 명령값 고스트
        if self._command is not None:
            cx = bar_x + bar_w * self._frac(self._command)
            p.setBrush(QColor(P.text_mute))
            p.drawRect(QRectF(cx - 1, y - 3, 2, 9))

        # 실제값. 한계 근접(양 끝 4%)은 마커 색으로만 알린다 —
        # 막대에 음영을 깔면 7줄이 겹쳐 얼룩처럼 보인다.
        ax = bar_x + bar_w * self._frac(self._actual)
        p.setBrush(QColor(P.warning) if self._near_limit() else QColor(P.accent))
        p.drawEllipse(QRectF(ax - 3.5, h / 2 - 3.5, 7, 7))

        p.setFont(_mono(9))
        p.setPen(QColor(P.text))
        p.drawText(QRectF(w - 52, 0, 52, h), Qt.AlignRight | Qt.AlignVCenter,
                   f"{math.degrees(self._actual):+.1f}°")
        p.end()

File: gcs/src/widgets/Gauges.cpp
#include "widgets/Gauges.h"

#include <cmath>
#include <QFont>
#include <QFontMetrics>
#include <QPainter>
#include <QPropertyAnimation>
#include <QtMath>

#include "theme/Tokens.h"

namespace gcs::ui {

using namespace gcs::theme;

namespace {

QFont monoFont(int pt, bool demiBold = false)
{
    QFont f = monoFont(10);
    f.setPointSize(pt);
    if (demiBold)
        f.setWeight(QFont::DemiBold);
    return f;
}

/// 게이지 트랙 색. 다크는 한 단계 밝은 면, 라이트는 한 단계 어두운 면을 쓴다.
QColor trackColor(const Colors &C)
{
    return QColor(C.isDark() ? C.surfaceHi : C.surfaceHover);
}

}  // namespace

// ============================ AnimatedValue ============================

AnimatedValue::AnimatedValue(QWidget *parent, int durationMs) : QWidget(parent)
{
    anim_ = new QPropertyAnimation(this, "value", this);
    anim_->setDuration(durationMs);
    anim_->setEasingCurve(QEasingCurve::OutCubic);
}

void AnimatedValue::setValue(double v)
{
    value_ = v;
    update();
}

void AnimatedValue::animateTo(double target)
{
    // 같은 값이 반복해서 들어올 때 애니메이션을 다시 시작하면 값이 영원히
    // 목표에 못 닿은 것처럼 보인다. 10 Hz 스트림에서는 흔한 상황이다.
    if (qAbs(target - value_) < 1e-4)
        return;
    anim_->stop();
    anim_->setStartValue(value_);
    anim_->setEndValue(target);
    anim_->start();
}

// ============================ BatteryRing ============================

BatteryRing::BatteryRing(QWidget *parent, int size, double lowThreshold)
    : AnimatedValue(parent), size_(size), low_(lowThreshold)
{
    setFixedSize(size_, size_);
}

void BatteryRing::setState(double socPercent, bool charging)
{
    charging_ = charging;
    animateTo(qBound(0.0, socPercent, 100.0));
}

void BatteryRing::paintEvent(QPaintEvent *)
{
    const Colors &C = colors();
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    const double thickness = 6.0;
    const double pad = thickness / 2 + 1;
    const QRectF rect(pad, pad, width() - pad * 2, height() - pad * 2);

    p.setPen(QPen(trackColor(C), thickness, Qt::SolidLine, Qt::FlatCap));
    p.drawArc(rect, 0, 360 * 16);

    QColor arc(C.success);
    if (charging_)
        arc = QColor(C.accent);
    else if (value() <= low_ * 0.6)
        arc = QColor(C.danger);
    else if (value() <= low_)
        arc = QColor(C.warning);

    if (value() > 0.3) {
        p.setPen(QPen(arc, thickness, Qt::SolidLine, Qt::FlatCap));
        // 12시에서 시계 방향으로 채운다.
        p.drawArc(rect, 90 * 16, -int(360 * 16 * value() / 100.0));
    }

    // 임계값 눈금 — 배경색으로 호를 끊어 표시한다.
    const double ang = qDegreesToRadians(90.0 - 360.0 * low_ / 100.0);
    const double cx = width() / 2.0;
    const double cy = height() / 2.0;
    const double rOut = rect.width() / 2 + thickness / 2;
    const double rIn = rect.width() / 2 - thickness / 2;
    p.setPen(QPen(QColor(C.surface), 2));
    p.drawLine(QPointF(cx + rIn * std::cos(ang), cy - rIn * std::sin(ang)),
               QPointF(cx + rOut * std::cos(ang), cy - rOut * std::sin(ang)));

    p.setFont(monoFont(int(size_ * 0.21), true));
    p.setPen(QColor(C.text));
    p.drawText(rect.adjusted(0, -5, 0, -5), Qt::AlignCenter,
               QString::number(value(), 'f', 0));

    QFont small;
    small.setPointSize(qMax(7, int(size_ * 0.09)));
    p.setFont(small);
    p.setPen(QColor(C.textMute));
    const int off = int(size_ * 0.27);
    p.drawText(rect.adjusted(0, off, 0, off), Qt::AlignCenter,
               charging_ ? QStringLiteral("충전 중") : QStringLiteral("%"));
}

// ============================ ArcGauge ============================

// ============================ BatteryPill ============================

BatteryPill::BatteryPill(QWidget *parent, double lowThreshold)
    : AnimatedValue(parent), low_(lowThreshold)
{
    setFixedSize(74, 22);
}

void BatteryPill::setState(double socPercent, bool charging)
{
    charging_ = charging;
    setToolTip(charging_ ? QStringLiteral("충전 중 · %1%").arg(socPercent, 0, 'f', 0)
                         : QStringLiteral("배터리 %1%").arg(socPercent, 0, 'f', 0));
    animateTo(qBound(0.0, socPercent, 100.0));
}

void BatteryPill::paintEvent(QPaintEvent *)
{
    const Colors &C = colors();
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    const double soc = value();
    const QColor tone = charging_       ? QColor(C.success)
                        : soc <= low_ / 2 ? QColor(C.danger)
                        : soc <= low_     ? QColor(C.warning)
                                          : QColor(C.success);

    // 셀 몸통과 오른쪽 단자.
    const QRectF cell(0.75, 4.5, 30.0, 13.0);
    p.setPen(QPen(QColor(C.borderHi), 1.2));
    p.setBrush(QColor(C.surfaceHi));
    p.drawRoundedRect(cell, 3, 3);

    p.setPen(Qt::NoPen);
    p.setBrush(QColor(C.borderHi));
    p.drawRoundedRect(QRectF(cell.right() + 1.0, 8.5, 2.4, 5.0), 1, 1);

    // 잔량. 0 % 에서도 실선이 보이도록 최소 폭을 준다.
    const double inner = cell.width() - 3.0;
    p.setBrush(tone);
    p.drawRoundedRect(QRectF(cell.left() + 1.5, cell.top() + 1.5,
                             qMax(2.0, inner * soc / 100.0), cell.height() - 3.0),
                      1.5, 1.5);

    QFont f;
    f.setWeight(QFont::DemiBold);
    p.setFont(f);
    p.setPen(QColor(soc <= low_ ? tone : QColor(C.text)));
    p.drawText(QRectF(38, 0, width() - 38, height()), Qt::AlignLeft | Qt::AlignVCenter,
               QStringLiteral("%1%").arg(soc, 0, 'f', 0));
}

ArcGauge::ArcGauge(QWidget *parent, const QString &caption, double warnBelow,
                   double dangerBelow)
    : AnimatedValue(parent), caption_(caption), warn_(warnBelow), danger_(dangerBelow)
{
    setMinimumHeight(70);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
}

void ArcGauge::setState(double normalized, const QString &rawText)
{
    rawText_ = rawText.isEmpty() ? QString::number(normalized, 'f', 3) : rawText;
    animateTo(qBound(0.0, normalized, 1.0));
}

void ArcGauge::paintEvent(QPaintEvent *)
{
    const Colors &C = colors();
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    const double thickness = 6.0;
    const double d = qMin(width() - 14.0, (height() - 22.0) * 2);
    const QRectF rect((width() - d) / 2, 9, d, d);

    p.setPen(QPen(trackColor(C), thickness, Qt::SolidLine, Qt::FlatCap));
    p.drawArc(rect, 180 * 16, -180 * 16);

    // 낮을수록 위험한 지표라 색이 역방향이다.
    QColor tone(C.success);
    if (value() <= danger_)
        tone = QColor(C.danger);
    else if (value() <= warn_)
        tone = QColor(C.warning);

    p.setPen(QPen(tone, thickness, Qt::SolidLine, Qt::FlatCap));
    p.drawArc(rect, 180 * 16, -int(180 * 16 * value()));

    p.setFont(monoFont(12, true));
    p.setPen(QColor(C.text));
    p.drawText(QRectF(0, height() - 32, width(), 18), Qt::AlignCenter, rawText_);

    if (!caption_.isEmpty()) {
        QFont f;
        f.setPointSize(8);
        p.setFont(f);
        p.setPen(QColor(C.textMute));
        p.drawText(QRectF(0, height() - 15, width(), 13), Qt::AlignCenter, caption_);
    }
}

// ============================ StatBar ============================

StatBar::StatBar(const QString &label, const QString &unit, QWidget *parent,
                 double warnAbove, double dangerAbove, double vmax)
    : QWidget(parent), label_(label), unit_(unit),
      warn_(warnAbove), danger_(dangerAbove), vmax_(vmax)
{
    setFixedHeight(26);
}

void StatBar::setReading(double v)
{
    value_ = v;
    valid_ = true;
    update();
}

void StatBar::paintEvent(QPaintEvent *)
{
    const Colors &C = colors();
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);
    const double w = width();

    QFont f;
    f.setPointSize(9);
    p.setFont(f);
    p.setPen(QColor(C.textMute));
    p.drawText(QRectF(0, 0, w * 0.55, 13), Qt::AlignLeft | Qt::AlignVCenter, label_);

    p.setFont(monoFont(9));
    p.setPen(valid_ ? QColor(C.text) : QColor(C.textMute));
    const QString txt = valid_ ? QString::number(value_, 'f', 0) + unit_
                               : QStringLiteral("—");
    p.drawText(QRectF(w * 0.45, 0, w * 0.55, 13), Qt::AlignRight | Qt::AlignVCenter, txt);

    p.setPen(Qt::NoPen);
    p.setBrush(trackColor(C));
    p.drawRect(QRectF(0, 18, w, 3));

    if (valid_ && value_ > 0) {
        QColor fill(C.textDim);
        if (danger_ >= 0 && value_ >= danger_)
            fill = QColor(C.danger);
        else if (warn_ >= 0 && value_ >= warn_)
            fill = QColor(C.warning);
        p.setBrush(fill);
        p.drawRect(QRectF(0, 18, qMax(2.0, w * qBound(0.0, value_ / vmax_, 1.0)), 3));
    }
}

// ============================ JointBar ============================

JointBar::JointBar(const QString &name, double lo, double hi, QWidget *parent)
    : QWidget(parent), name_(name), lo_(lo), hi_(hi)
{
    setFixedHeight(22);
}

void JointBar::setActual(double rad)
{
    actual_ = rad;
    update();
}

void JointBar::setCommand(double rad)
{
    command_ = rad;
    hasCommand_ = true;
    update();
}

void JointBar::clearCommand()
{
    hasCommand_ = false;
    update();
}

double JointBar::fraction(double v) const
{
    return qBound(0.0, (v - lo_) / (hi_ - lo_), 1.0);
}

bool JointBar::nearLimit() const
{
    const double f = fraction(actual_);
    return f < 0.04 || f > 0.96;
}

void JointBar::paintEvent(QPaintEvent *)
{
    const Colors &C = colors();
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    const double w = width();
    const double h = height();
    const double barX = 26;
    const double barW = w - barX - 56;

    QFont f;
    f.setPointSize(9);
    p.setFont(f);
    p.setPen(QColor(C.textMute));
    p.drawText(QRectF(0, 0, 22, h), Qt::AlignLeft | Qt::AlignVCenter, name_);

    const double y = h / 2 - 1.5;
    p.setPen(Qt::NoPen);
    p.setBrush(trackColor(C));
    p.drawRect(QRectF(barX, y, barW, 3));

    // 명령값 고스트. 실제값과 갈라지는 순간이 계획 실패·리플렉스의 신호다.
    if (hasCommand_) {
        const double cx = barX + barW * fraction(command_);
        p.setBrush(QColor(C.textMute));
        p.drawRect(QRectF(cx - 1, y - 3, 2, 9));
    }

    // 실제값. 한계 근접은 마커 색으로만 알린다 — 막대에 음영을 깔면
    // 7 축이 세로로 쌓였을 때 얼룩처럼 보인다.
    const double ax = barX + barW * fraction(actual_);
    p.setBrush(nearLimit() ? QColor(C.warning) : QColor(C.accent));
    p.drawEllipse(QRectF(ax - 3.5, h / 2 - 3.5, 7, 7));

    p.setFont(monoFont(9));
    p.setPen(QColor(C.text));
    p.drawText(QRectF(w - 52, 0, 52, h), Qt::AlignRight | Qt::AlignVCenter,
               QStringLiteral("%1°").arg(qRadiansToDegrees(actual_), 0, 'f', 1));
}

}  // namespace gcs::ui
